File: sub_projs/cllm/tools/llm-login/oauth.cpp
// oauth.cpp — OAuth 2.0 授權碼＋PKCE 實作。出站重用 cllm 的 http（src/http）。

#include "oauth.hpp"

#include <stdexcept>

#include "http.hpp"    // cllm 出站笨管子
#include "httpd.hpp"   // 共用入站 server（接 callback）

namespace login::oauth {
namespace oauth_impl {  // 具名子 namespace（對齊專案慣例）

std::string sget(J &j, const char *k, std::string d = "") {
  return j.is_object() && j.contains(k) && j[k].is_string() ? j[k].get_string() : d;
}

int iget(J &j, const char *k, int d) {
  return j.is_object() && j.contains(k) && j[k].is_number()
             ? static_cast<int>(j[k].get_number())
             : d;
}

std::string url_encode(const std::string &s) {
  static const char *hex = "0123456789ABCDEF";
  std::string out;
  for (unsigned char c : s) {
    if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') ||
        c == '-' || c == '_' || c == '.' || c == '~') {
      out.push_back(static_cast<char>(c));
    } else {
      out.push_back('%');
      out.push_back(hex[c >> 4]);
      out.push_back(hex[c & 15]);
    }
  }
  return out;
}

std::string append_param(std::string url, const std::string &k, const std::string &v) {
  url.push_back(url.find('?') == std::string::npos ? '?' : '&');
  url += url_encode(k) + "=" + url_encode(v);
  return url;
}

// POST 一包，解回應 JSON；後端錯誤（4xx 或含 error）→ throw 帶訊息。
J post_json_reply(const std::string &url, const std::vector<std::string> &headers,
                  const std::string &body) {
  http::Request req;
  req.url = url;
  req.method = "POST";
  req.headers = headers;
  req.body = body;
  http::Response resp = http::request(req);
  J j;
  bool ok = !glz::read_json(j, resp.body);
  if (resp.status >= 400 || (ok && j.is_object() && j.contains("error"))) {
    std::string msg = ok && j.is_object() && j.contains("error")
                          ? (j["error"].is_string() ? j["error"].get_string()
                                                    : resp.body.substr(0, 200))
                          : resp.body.substr(0, 200);
    throw std::runtime_error("token 端錯誤 (HTTP " + std::to_string(resp.status) + "): " + msg);
  }
  if (!ok) throw std::runtime_error("token 回應非 JSON：" + resp.body.substr(0, 200));
  return j;
}

}  // namespace oauth_impl
using namespace oauth_impl;

Redirect redirect_of(J &prov) {
  Redirect r;
  r.host = sget(prov, "redirect_host", "127.0.0.1");
  r.port = iget(prov, "redirect_port", 8123);
  r.path = sget(prov, "redirect_path", "/callback");
  r.uri = "http://" + r.host + ":" + std::to_string(r.port) + r.path;
  return r;
}

std::string authorize_url(J &prov, const std::string &redirect_uri, const std::string &state,
                          const std::string &challenge) {
  std::string url = sget(prov, "authorize_url");
  url = append_param(url, "response_type", "code");
  std::string cid = sget(prov, "client_id");
  if (!cid.empty()) url = append_param(url, "client_id", cid);
  url = append_param(url, "redirect_uri", redirect_uri);
  std::string scope = sget(prov, "scope");
  if (!scope.empty()) url = append_param(url, "scope", scope);
  if (!state.empty()) url = append_param(url, "state", state);
  url = append_param(url, "code_challenge", challenge);
  url = append_param(url, "code_challenge_method", "S256");
  return url;
}

std::string authorize_url_openrouter(J &prov, const std::string &redirect_uri,
                                     const std::string &challenge) {
  std::string url = sget(prov, "authorize_url", "https://openrouter.ai/auth");
  url = append_param(url, "callback_url", redirect_uri);
  url = append_param(url, "code_challenge", challenge);
  url = append_param(url, "code_challenge_method", "S256");
  return url;
}

std::string capture_code(const std::string &host, int port, const std::string &path,
                         const std::string &expected_state) {
  std::string code, err;
  httpd::Server srv(host, port);
  srv.serve_once([&](const httpd::Request &req, httpd::Responder &res) {
    if (req.path != path) {
      res.send(404, "text/plain", "not found");
      return;
    }
    std::string got_state = req.query_param("state");
    code = req.query_param("code");
    if (code.empty()) {
      err = "callback 沒帶 code（" + req.query + "）";
      res.send(400, "text/html", "<h3>登入失敗：沒拿到 code，可關閉本頁。</h3>");
      return;
    }
    if (!expected_state.empty() && got_state != expected_state) {
      err = "state 不符（可能 CSRF）";
      code.clear();
      res.send(400, "text/html", "<h3>登入失敗：state 不符，可關閉本頁。</h3>");
      return;
    }
    res.send(200, "text/html", "<h3>登入完成，可關閉本頁回終端機。</h3>");
  });
  if (!err.empty()) throw std::runtime_error(err);
  if (code.empty()) throw std::runtime_error("沒拿到授權碼");
  return code;
}

J exchange_code(J &prov, const std::string &code, const std::string &redirect_uri,
                const std::string &verifier) {
  std::string body = "grant_type=authorization_code";
  body += "&code=" + url_encode(code);
  body += "&redirect_uri=" + url_encode(redirect_uri);
  body += "&code_verifier=" + url_encode(verifier);
  std::string cid = sget(prov, "client_id");
  if (!cid.empty()) body += "&client_id=" + url_encode(cid);
  std::string secret = sget(prov, "client_secret");
  if (!secret.empty()) body += "&client_secret=" + url_encode(secret);
  return post_json_reply(sget(prov, "token_url"),
                         {"content-type: application/x-www-form-urlencoded"}, body);
}

J exchange_code_openrouter(J &prov, const std::string &code, const std::string &verifier) {
  std::string url = sget(prov, "token_url", "https://openrouter.ai/api/v1/auth/keys");
  std::string body = "{\"code\":\"" + code + "\",\"code_verifier\":\"" + verifier +
                     "\",\"code_challenge_method\":\"S256\"}";
  return post_json_reply(url, {"content-type: application/json"}, body);
}

J refresh(J &prov, const std::string &refresh_token) {
  std::string body = "grant_type=refresh_token";
  body += "&refresh_token=" + url_encode(refresh_token);
  std::string cid = sget(prov, "client_id");
  if (!cid.empty()) body += "&client_id=" + url_encode(cid);
  std::string secret = sget(prov, "client_secret");
  if (!secret.empty()) body += "&client_secret=" + url_encode(secret);
  return post_json_reply(sget(prov, "token_url"),
                         {"content-type: application/x-www-form-urlencoded"}, body);
}

}  // namespace login::oauth

File: sub_projs/cllm/tools/llm-login/oauth.py
"""oauth.py — OAuth 2.0 授權碼＋PKCE 協定層（供應商中立、零外部相依）。

只做「協定」：產 PKCE、組 authorize URL、起本機 callback 接授權碼、拿碼換 token、
refresh。所有供應商差異（authorize_url／token_url／client_id／scopes…）都由外部
provider dict 餵進來，本檔不硬編任何一家。純標準庫（urllib／http.server／hashlib…）。
"""

from __future__ import annotations

import base64
import hashlib
import http.server
import json
import secrets
import threading
import urllib.parse
import urllib.request


# ── PKCE（RFC 7636）───────────────────────────────────────────────
def gen_pkce() -> tuple[str, str]:
    """回傳 (verifier, challenge)。challenge = base64url(sha256(verifier))，無填充。"""
    verifier = secrets.token_urlsafe(64)  # >43 字元，符合 RFC 長度下限
    digest = hashlib.sha256(verifier.encode("ascii")).digest()
    challenge = base64.urlsafe_b64encode(digest).rstrip(b"=").decode("ascii")
    return verifier, challenge


# ── 組 authorize URL ─────────────────────────────────────────────
def build_authorize_url(provider: dict, redirect_uri: str,
                        state: str, challenge: str) -> str:
    """把 authorize 端點＋標準參數＋供應商自訂 extra_auth_params 組成完整 URL。"""
    params = {
        "response_type": "code",
        "client_id": provider["client_id"],
        "redirect_uri": redirect_uri,
        "state": state,
        "code_challenge": challenge,
        "code_challenge_method": "S256",
    }
    scopes = provider.get("scopes")
    if scopes:
        params["scope"] = " ".join(scopes) if isinstance(scopes, list) else scopes
    params.update(provider.get("extra_auth_params", {}))
    return provider["authorize_url"] + "?" + urllib.parse.urlencode(params)


# ── 本機 callback：接一次授權碼就關 ───────────────────────────────
class _Catcher(http.server.BaseHTTPRequestHandler):
    result: dict = {}
    done = threading.Event()

    def do_GET(self):  # noqa: N802（http.server 介面命名）
        q = urllib.parse.urlparse(self.path).query
        _Catcher.result = dict(urllib.parse.parse_qsl(q))
        self.send_response(200)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.end_headers()
        msg = "登入完成，可以關掉這個分頁回到終端機。" if "code" in _Catcher.result \
            else "授權失敗：" + _Catcher.result.get("error", "未知錯誤")
        self.wfile.write(("<html><body><h3>%s</h3></body></html>" % msg).encode("utf-8"))
        _Catcher.done.set()

    def log_message(self, *_):  # 靜音：不要把每個 request 印到 stderr
        pass


def capture_code(host: str, port: int, expect_state: str, timeout_s: int = 300) -> str:
    """阻塞直到 callback 進來；驗 state、回授權碼。state 不符或帶 error 即拋。"""
    _Catcher.result = {}
    _Catcher.done.clear()
    srv = http.server.HTTPServer((host, port), _Catcher)
    threading.Thread(target=srv.serve_forever, daemon=True).start()
    try:
        if not _Catcher.done.wait(timeout_s):
            raise TimeoutError("等 callback 逾時（%ds）——授權沒完成。" % timeout_s)
    finally:
        srv.shutdown()
        srv.server_close()
    res = _Catcher.result
    if "error" in res:
        raise RuntimeError("授權端回錯：%s（%s）"
                          % (res.get("error"), res.get("error_description", "")))
    if res.get("state") != expect_state:
        raise RuntimeError("state 不符——可能遭 CSRF，中止。")
    if "code" not in res:
        raise RuntimeError("callback 沒帶授權碼。")
    return res["code"]


# ── token 端點：換碼／refresh ─────────────────────────────────────
def _post_token(provider: dict, form: dict) -> dict:
    """POST application/x-www-form-urlencoded 到 token_url，回解析後的 JSON dict。"""
    secret = provider.get("client_secret")
    if secret:
        form["client_secret"] = secret
    data = urllib.parse.urlencode(form).encode("ascii")
    req = urllib.request.Request(provider["token_url"], data=data, method="POST")
    req.add_header("Content-Type", "application/x-www-form-urlencoded")
    req.add_header("Accept", "application/json")
    with urllib.request.urlopen(req, timeout=provider.get("timeout_s", 60)) as resp:
        return json.loads(resp.read().decode("utf-8"))


def exchange_code(provider: dict, code: str, redirect_uri: str, verifier: str) -> dict:
    """授權碼 → token 集（access／refresh／expires_in…）。"""
    return _post_token(provider, {
        "grant_type": "authorization_code",
        "code": code,
        "redirect_uri": redirect_uri,
        "client_id": provider["client_id"],
        "code_verifier": verifier,
    })


def refresh(provider: dict, refresh_token: str) -> dict:
    """refresh_token → 新的 token 集。有些供應商會輪換 refresh_token，呼叫端要存回。"""
    return _post_token(provider, {
        "grant_type": "refresh_token",
        "refresh_token": refresh_token,
        "client_id": provider["client_id"],
    })
